Update `rand` in the stdlib tests, and remove the `getrandom` feature from it.

thomcc opened the pull request:
The main goal is actually removing `getrandom`, so that eventually we can allow running the stdlib test suite on tier3 targets which don't have `getrandom` support. Currently those targets can only run the subset of stdlib tests that exist in uitests, and (generally speaking), we prefer not to test libstd functionality in uitests, which came up recently in https://github.com/rust-lang/rust/pull/104095 and https://github.com/rust-lang/rust/pull/104185. Additionally, the fact that we can't update `rand`/`getrandom` means we're stuck with the old set of tier3 targets, so can't test new ones.

~~Anyway, I haven't checked that this actually does allow use on tier3 targets (I think it does not, as some work is needed in stdlib submodules) but it moves us slightly closer to this, and seems to allow at least finally updating our `rand` dep, which definitely improves the status quo.~~ Checked and works now.

For the most part, our tests and benchmarks are fine using hard-coded seeds. A couple tests seem to fail with this (stuff manipulating the environment expecting no collisions, for example), or become pointless (all inputs to a function become equivalent). In these cases I've done a (gross) dance (ab)using `RandomState` and `Location::caller()` for some extra "entropy".

Trying to share that code seems *way* more painful than it's worth given that the duplication is a 7-line function, even if the lines are quite gross. (Keeping in mind that sharing it would require adding `rand` as a non-dev dep to std, and exposing a type from it publicly, all of which sounds truly awful, even if done behind a perma-unstable feature).

See also some previous attempts:
- https://github.com/rust-lang/rust/pull/86963 (in particular https://github.com/rust-lang/rust/pull/86963#issuecomment-885438936 which explains why this is non-trivial)
- https://github.com/rust-lang/rust/pull/89131
- https://github.com/rust-lang/rust/pull/96626#issuecomment-1114562857 (I tried in that PR at the same time, but settled for just removing the usage of `thread_rng()` from the benchmarks, since that was the main goal).
- https://github.com/rust-lang/rust/pull/104185
- Probably more. It's very tempting of a thing to "just update".

r? @Mark-Simulacrum

thomcc committed:
Update rand in the stdlib tests, and remove the getrandom feature from it

thomcc commented:
Is the SmallRng seed's size really target-dependent? That's awful.

thomcc commented:
I'll look into that and finish this up tomorrow.

@rustbot author

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3510606772/jobs/5880564852) [(plain)](https://github.com/rust-lang/rust/commit/a70046d7a7c5a0fadc0afa8e87f841a4d0a105b5/checks/9605676278/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
    Checking rand_xorshift v0.3.0
    Checking std v0.0.0 (/checkout/library/std)
    Checking alloc v0.0.0 (/checkout/library/alloc)
    Checking core v0.0.0 (/checkout/library/core)
error[E0271]: type mismatch resolving `<SmallRng as SeedableRng>::Seed == [u8; 32]`
   |
   |
16 |     rand::SeedableRng::from_seed(seed)
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an array with a fixed size of 16 elements, found one with 32 elements
For more information about this error, try `rustc --explain E0271`.
error: could not compile `std` due to previous error
warning: build failed, waiting for other jobs to finish...
warning: build failed, waiting for other jobs to finish...
error[E0271]: type mismatch resolving `<SmallRng as SeedableRng>::Seed == [{integer}; 32]`
  --> library/alloc/benches/lib.rs:25:5
   |
25 | /     rand::SeedableRng::from_seed([
26 | |         0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
27 | |         25, 26, 27, 28, 29, 30, 31,
28 | |     ])
   | |______^ expected an array with a fixed size of 16 elements, found one with 32 elements
error: could not compile `alloc` due to previous error
error: could not compile `alloc` due to previous error
error[E0271]: type mismatch resolving `<SmallRng as SeedableRng>::Seed == [{integer}; 32]`
   --> library/alloc/src/collections/binary_heap/tests.rs:443:41
    |
443 |       let mut rng: rand::rngs::SmallRng = rand::SeedableRng::from_seed([
444 | |         0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
445 | |         25, 26, 27, 28, 29, 30, 31,
446 | |     ]);
446 | |     ]);
    | |______^ expected an array with a fixed size of 16 elements, found one with 32 elements

error[E0271]: type mismatch resolving `<SmallRng as SeedableRng>::Seed == [{integer}; 32]`
   --> library/alloc/src/collections/linked_list/tests.rs:523:41
    |
523 |       let mut rng: rand::rngs::SmallRng = rand::SeedableRng::from_seed([
524 | |         0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
525 | |         25, 26, 27, 28, 29, 30, 31,
526 | |     ]);
526 | |     ]);
    | |______^ expected an array with a fixed size of 16 elements, found one with 32 elements

error[E0271]: type mismatch resolving `<SmallRng as SeedableRng>::Seed == [{integer}; 32]`
   |
   |
96 | /     rand::SeedableRng::from_seed([
97 | |         0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24,
98 | |         25, 26, 27, 28, 29, 30, 31,
99 | |     ])
   | |______^ expected an array with a fixed size of 16 elements, found one with 32 elements
error: could not compile `alloc` due to previous error
error: could not compile `alloc` due to 2 previous errors
error: could not compile `alloc` due to 2 previous errors
error[E0271]: type mismatch resolving `<SmallRng as SeedableRng>::Seed == [u8; 32]`
    |
    |
166 |     rand::SeedableRng::from_seed(seed)
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an array with a fixed size of 16 elements, found one with 32 elements
error: could not compile `core` due to previous error
error: could not compile `core` due to previous error
error[E0271]: type mismatch resolving `<SmallRng as SeedableRng>::Seed == [u8; 32]`
    |
    |
675 |         rand::SeedableRng::from_seed(seed)
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected an array with a fixed size of 16 elements, found one with 32 elements
error: could not compile `std` due to previous error
Build completed unsuccessfully in 0:00:37
```

</details>

rust-log-analyzer commented:
The job **`dist-i586-gnu-i586-i686-musl`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3512104317/jobs/5883480341) [(plain)](https://github.com/rust-lang/rust/commit/392b168197a1eb5eb764499b180f5c6868c9a77b/checks/9609247797/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
  IMAGE: dist-i586-gnu-i586-i686-musl
  AWS_ACCESS_KEY_ID: 
  AWS_SECRET_ACCESS_KEY: 
##[endgroup]
cp: cannot stat 'obj/build/metrics.json': No such file or directory
##[error]Process completed with exit code 1.
```

</details>

thomcc commented:
Ah, my bad, some of those are not tests.

thomcc commented:
This [seems](https://github.com/rust-lang/rust/actions/runs/3512386127) to pass on all the targets I care to test on that would normally fail (most notably [wasm32](https://github.com/rust-lang/rust/actions/runs/3512386127/jobs/5884035472)), which implies it really has excised `getrandom` from the stdlib's dev-dep tree (this seems to be confirmed by `cargo tree`, but that can be misleading).

So I don't think anything else is needed here, and it's ready for review.

@rustbot ready

RalfJung commented on library/std/tests/env.rs:
```diff
@@ -2,11 +2,27 @@ use std::env::*;
+    use rand::Rng;
+    let n = format!(
+        "TEST{}",
+        test_rng().sample_iter(&Alphanumeric).map(|byte| byte as char).take(10).collect::<String>(),
```
I think `sample_string` is the function meant to be used for this now

RalfJung commented on library/alloc/benches/slice.rs:
```diff
@@ -218,7 +218,7 @@ fn gen_strings(len: usize) -> Vec<String> {
     for _ in 0..len {
         let n = rng.gen::<usize>() % 20 + 1;
-        v.push((&mut rng).sample_iter(&Alphanumeric).take(n).collect());
+        v.push((&mut rng).sample_iter(&Alphanumeric).map(|byte| byte as char).take(n).collect());
```
I think `sample_string` is the function meant to be used for this now

thomcc replied:
Thanks, done.

klensy commented:
If all goes well, maybe bump getrandom (https://github.com/rust-random/getrandom/blob/master/CHANGELOG.md#028---2022-10-20) too?

thomcc commented:
I'll look into bumping `getrandom` in a later PR when I can examine the changelog more closely (which is something we try to do for non-test-only deps, even well-vetted ones like `getrandom`). Part of my short-term motivation is unblocking my testing of https://github.com/rust-lang/rust/pull/103503, which does require https://github.com/rust-random/getrandom/pull/317, and thus a newish getrandom (although I'm unsure if that's been in a release yet).

the8472 commented:
> For the most part, our tests and benchmarks are fine using hard-coded seeds. A couple tests seem to fail with this (stuff manipulating the environment expecting no collisions, for example), or become pointless (all inputs to a function become equivalent). In these cases I've done a (gross) dance (ab)using RandomState and Location::caller() for some extra "entropy".

It feels iffy to run randomized tests with fixed seeds. Can we conditionally import and use the getrandom feature on some tier-1 targets?

thomcc commented:
> It feels iffy to run randomized tests with fixed seeds. Can we conditionally import and use the getrandom feature on some tier-1 targets?

Doing this conditionally would be more painful since I'd have to duplicate a body of a `#[cfg(...)]` and a `[target.'cfg(...)'.dev-dependencies]`  in more places. It also would add more conditional compliation that only fails on obscure targets, which is annoying for maintenance. I'm not sure it's really worth it given that most of those cases I've tried to use a RNG method that isn't a fixed seed (e.g. the hacky `test_rng()` thing should genuinely give a different seed at least each test run).

I guess I'll take a look at what it looks like though.

the8472 commented on library/std/tests/env.rs:
```diff
@@ -1,12 +1,24 @@
+#[track_caller]
+pub(crate) fn test_rng() -> rand_xorshift::XorShiftRng {
+    use core::hash::{BuildHasher, Hash, Hasher};
+    let mut hasher = std::collections::hash_map::RandomState::new().build_hasher();
```
Wait, this means we already pull in non-determinism on platforms that have have a randomness source, right? Then my previous concern is moot for std tests.

But alloc tests still need a similar solution.

thomcc replied:
I'll just use this there. It's gross but I'm doing the `cfg` approach now and it's also ending up really gross, and wouldn't remove this grossness.

thomcc replied:
I've removed all use of hard-coded seeds from the tests (we still use them in benches, which is intentional, since nondeterminism is bad in those).

This was kind of a pain in the ass, but it would have been painful anyway I guess.

thomcc commented:
> I guess I'll take a look at what it looks like though.

As per https://github.com/rust-lang/rust/pull/104658#discussion_r1028418799 I didn't end up going this route, but I have made it so that the places that were using non-determinism to improve testing are still using it.

thomcc commented on library/alloc/src/slice/tests.rs:
```diff
@@ -0,0 +1,359 @@
+use crate::borrow::ToOwned;
```
This whole file is the subset of `alloc/test/slice.rs` that uses non-determinism. There are clearly some things in it than need cleaning up, but I'd rather do that at a future time.

the8472 replied:
Making the test helper public would have worked too. Since it's gated on cfg(test) it would only be available here but not to crate users.

thomcc replied:
Hm? `cfg(test)` isn't enabled for integration tests, only unit tests.

the8472 replied:
Huh? I thought I had seen something like that before. Maybe it was through a perma-unstable feature instead. But nevermind, if moving the code is the simplest option, so be it.

thomcc replied:
Yeah I'd definitely prefer copypaste to exposing a type from an external crate in std. Even as perma-unstable. One option to avoid the duplication would be `include!`ing a file, though. But I don't know if that's better. If the reviewer would prefer it I don't mind, but IDK where the file would go (and also it needs to be slightly different for std vs everything else, since `std::blah` doesn't work in std)

flba-eb commented:
I've tested this branch with the WIP patch for QNX/Neutrino: This patch allows us to run stdlib tests without having to modify `getrandom`. Thanks @thomcc!!

bors commented:
:umbrella: The latest upstream changes (presumably #103693) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Mark-Simulacrum commented:
I'm broadly fine with this, but it would make me feel better if we had some brief discussion at a T-libs meeting (I don't need to be there) to confirm we're OK with the broad strokes proposed here (i.e., we don't believe that we need a "better" randomness source).

I'd also like to see one of the sites (probably the libstd one) annotated with some details on where we're pulling randomness from. Maybe worth doing something like hashing a pointer value so we can get entropy from address-space layout randomization, though I'd guess some tier 3 platforms don't really have that.

thomcc commented:
> I'm broadly fine with this, but it would make me feel better if we had some brief discussion at a T-libs meeting

I missed this comment. Sounds good to me.

thomcc commented on library/std/src/lib.rs:
```diff
@@ -652,3 +652,30 @@ mod sealed {
+    #[track_caller]
+    pub(crate) fn test_rng() -> rand_xorshift::XorShiftRng {
+        use core::hash::{BuildHasher, Hash, Hasher};
+        let mut hasher = crate::collections::hash_map::RandomState::new().build_hasher();
```
Yes, RandomState is not actually random on all platforms.

bors commented:
:umbrella: The latest upstream changes (presumably #106033) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

bjorn3 commented on library/core/Cargo.toml:
```diff
@@ -24,8 +24,8 @@ path = "benches/lib.rs"
-rand = "0.7"
-rand_xorshift = "0.2"
+rand = { version = "0.8.5", default-features = false }
+rand_xorshift = { version = "0.3.0", default-features = false }
```
Note to self: update https://github.com/bjorn3/rustc_codegen_cranelift/blob/master/patches/0022-sysroot-Disable-not-compiling-tests.patch

thomcc commented:
This was discussed in the most recent t-libs meeting. The [consensus](https://rust-lang.zulipchat.com/#narrow/stream/259402-t-libs.2Fmeetings/topic/Meeting.202023-01-04/near/319422410) was that this implementation is acceptable.

I've rebased and squashed (to simplify the rebase) so this should be good to go.

@rustbot ready

Mark-Simulacrum commented:
@bors r+ rollup=iffy

bors commented:
:pushpin: Commit a4bf36e87bdec61240fb3040774d008c70acbfbb has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum a4bf36e87bdec61240fb3040774d008c70acbfbb -->
<!-- homu: {"type":"Approved","sha":"a4bf36e87bdec61240fb3040774d008c70acbfbb","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit a4bf36e87bdec61240fb3040774d008c70acbfbb with merge 2afe58571e53d48a1fc2354271abe5aff60c5c44...
<!-- homu: {"type":"BuildStarted","head_sha":"a4bf36e87bdec61240fb3040774d008c70acbfbb","merge_sha":"2afe58571e53d48a1fc2354271abe5aff60c5c44"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3864831561/jobs/6588917980)
Approved by: Mark-Simulacrum
Pushing 2afe58571e53d48a1fc2354271abe5aff60c5c44 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3864831561/jobs/6588917980"},"merge_sha":"2afe58571e53d48a1fc2354271abe5aff60c5c44"} -->

rust-timer commented:
Finished benchmarking commit (2afe58571e53d48a1fc2354271abe5aff60c5c44): [comparison URL](https://perf.rust-lang.org/compare.html?start=e5d46a5bdac3f19793297914d4ada432024fbe95&end=2afe58571e53d48a1fc2354271abe5aff60c5c44&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=e5d46a5bdac3f19793297914d4ada432024fbe95&end=2afe58571e53d48a1fc2354271abe5aff60c5c44&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | 0.7%  | [0.5%, 0.8%]   | 7     |
| Improvements ✅ <br /> (primary)   | -     | -              | 0     |
| Improvements ✅ <br /> (secondary) | -0.1% | [-0.1%, -0.1%] | 1     |
| All ❌✅ (primary)                 | -     | -              | 0     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=e5d46a5bdac3f19793297914d4ada432024fbe95&end=2afe58571e53d48a1fc2354271abe5aff60c5c44&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=e5d46a5bdac3f19793297914d4ada432024fbe95&end=2afe58571e53d48a1fc2354271abe5aff60c5c44&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.
